crates/edr_napi/src/solidity_tests/runner.rs

```diff
@@ -0,0 +1,147 @@
+/// Based on `crates/foundry/forge/tests/it/test_helpers.rs`.
+use std::{path::PathBuf, sync::Arc};
+
+use alloy_primitives::U256;
+use forge::{
+    constants::CALLER,
+    decode::RevertDecoder,
+    multi_runner::TestContract,
+    opts::{Env as EvmEnv, EvmOpts},
+    revm::primitives::SpecId,
+    MultiContractRunner, MultiContractRunnerBuilder, TestOptions, TestOptionsBuilder,
+};
+use foundry_compilers::{ArtifactId, ProjectCompileOutput};
+use foundry_config::{
+    Config, FuzzConfig, FuzzDictionaryConfig, InvariantConfig, RpcEndpoint, RpcEndpoints,
+};
+
+pub(super) fn build_runner(
+    test_suites: Vec<(ArtifactId, TestContract)>,
+) -> napi::Result<MultiContractRunner> {
+    let config = foundry_config();
+    let mut evm_opts = evm_opts();
+    evm_opts.isolate = config.isolate;
+
+    let project = config
+        .project()
+        .map_err(|e| napi::Error::new(napi::Status::GenericFailure, e.to_string()))?;
+    let compiled = project
+        .compile()
+        .map_err(|e| napi::Error::new(napi::Status::GenericFailure, e.to_string()))?;
+
+    let builder = MultiContractRunnerBuilder::new(Arc::new(config))
+        .sender(evm_opts.sender)
+        .with_test_options(test_opts(&compiled));
+
+    let abis = test_suites.iter().map(|(_, contract)| &contract.abi);
+    let revert_decoder = RevertDecoder::new().with_abis(abis);
+
+    let evm_env = evm_opts.local_evm_env();
+
+    Ok(MultiContractRunner {
+        contracts: test_suites.into_iter().collect(),
+        evm_opts,
+        env: evm_env,
+        evm_spec: builder.evm_spec.unwrap_or(SpecId::MERGE),
+        sender: builder.sender,
+        revert_decoder,
+        fork: builder.fork,
+        config: builder.config,
+        coverage: builder.coverage,
+        debug: builder.debug,
+        test_options: builder.test_options.unwrap_or_default(),
+        isolation: builder.isolation,
+        output: compiled,
+    })
+}
+
+fn project_root() -> PathBuf {
+    PathBuf::from(concat!(env!("CARGO_MANIFEST_DIR"), "/../../testdata"))
```

Comment:
I'm trying to wrap my head around this. My understanding is:

`CARGO_MANIFEST_DIR` = `./crates/edr_napi/`
Concatenated, you get: `./crates/edr_napi/../../testdata`
which equals: `./testdata`

However, `testdata` is found at `crates/foundry/testdata`.

What am I missing here? 😅

Reply:
Thanks good catch the path was wrong! Fixed in [bf5cb47](https://github.com/NomicFoundation/edr/pull/474/commits/bf5cb47d15d1962f214e9e728557603032cd2595)

It didn't cause failures, because the contracts are passed in through JS, not loaded from disk